---
title: "MADRID: Methylation Age Determination via Read-Informated Data"
subtitle: "Andy Madrid"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{introduction}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## General Usage
This is a short introduction to the general functionality of the madrid package. In short, this package takes a bsseq object, filters CpGs to only those select few used to determine epigenetic age, impute any missing values, and then predicts age for each sample.


There are a few more functions, such as guessing the sex of samples, but those are very much in the beta testing stage.


Let's start by installing some packages that are needed to install and run the madrid package.

## Installation
```{r}
### Install required packages
# install.packages(c("devtools","bsseq","glmnet", "mice", "wateRmelon", "DunedinPACE"))


### install the madrid package
# install_github("andymadrid/madrid")


```

Now we can load the madrid package.

## Setup
```{r setup}
# load in the madrid pacakge
suppressPackageStartupMessages(library(madrid))


```
Now we can load in a toy dataset.

```{r}
# get toy example
data(bs)


```
Let us take a look at this toy example.

```{r}
# summary of bsseq object
bs

# summary of phenotypic data associated with bsseq object
pData(bs)


```
We see we have 10 samples. Due to size limitations, this bsseq object only has the CpGs used to predict age, along with CpGs from the sex chromosomes. However, for your purposes, there is no need to remove any CpGs from your bsseq object. In fact, that is not advised.

The bsseq object also has some phenotypic data for the samples, which includes the age, sex, and (pseudo) diagnostic group for each.


## Guess sex of samples

Now we can attempt to guess the sex of the samples.

```{r}
# guess sample sex from bsseq object
guessed_sex <- guessSex(bs)

# look at the output
guessed_sex


```
We see that the guessed sex matches up nicely with the actual sex of the samples.

## Predict age

Now lets predict the age of the samples
```{r}
# predict age from bsseq object
predicted_age <- predictAge(bs)

# look at the output
predicted_age


```
This outputs the predicted ages for each sample. We can also see how well they correlate with the actual ages, if that information is readily available.

Please note, by default the predictAge() function will estimate age using array-based clocks, as well as the sequencing-based clock. To do this, it leans heavily on two other packages, wateRmelon and DunedinPACE. At its core, for this portion of the function, it really just formats the methylation matrix accordingly and filters CpGs to only those on the arrays. In this toy example most of the ages look off due to the fact that most CpGs are missing because of size limitations. That being said, in your own data, estimates should (hopefully) be more accurate.
```{r}
# assess the correlation between actual and estimated age
corMADRID <- cor(predicted_age$MADRID_Age, pData(bs)$Age)
corMADRID

```
We see that we have a fairly nice correlation between predicted and actual age. And that's really all that there is to it. You can take these results and see if there are any signficiant differences between the estimated and actual age of samples between groups.

```{r}
# test for differences (accelerated/decelerated) in actual and estimated age
predicted_age$difference <- predicted_age$MADRID_Age - pData(bs)$Age

t.test(predicted_age$difference ~ factor(pData(bs)$Diagnosis))


```
Some people prefer to run the test on the residuals, rather than on the direct differences in age themselves. That would look something like this...

```{r}
# test for differences in the resduals between actual and estimated age
predicted_age$residuals <- resid(lm(predicted_age$MADRID_Age ~ pData(bs)$Age))
t.test(predicted_age$residuals ~ factor(pData(bs)$Diagnosis))


```

I'd like to note that a Wilcoxon test or something like a generalized linear model that adjusts for known covariates would probably be more approporaite test to use, but for the sake of demonstration a simple t-test is utilized here.

That being said, in this toy example there is no signficant difference (accelerated or decelerated) in predicted age between groups. Hopefully in your data there will be, though!



If you're using blood samples, we can also estiamte the cell-type proportions from your sequenced data. The function estimateCellCounts() leans heavily upon the R package meffil to perform these calculations. In testing, I had some samples that were run on the EPIC microarray and were also sequenced and found that this method produced results that were more correlated with each other than other methods that were developed for sequence data (e.g., methylcc).


``` {r}
# estimate cell-type proportions from blood samples
# cellProps <- estimateCellCounts(bs)

# check the estimates
# head(cellProps)

```

Please note, the above code is not actually run since the bsseq object we have here for the demonstration does not have the sufficient CpGs needed estimate cell-type proportions and would throw out an error. But, assuming your bsseq object has all the CpGs, that shouldn't be an issue for you!


While epigenetic age is the main function of this package, I've also added in a a function to identify stochastic epigenetic mutations (SEMs). For more information on SEMs please refer to Gentilini et al (2015) Aging and Markov et al (2024) GeroScience.

Essentially, this function takes your data, identifies the interquartile range (IQR) for each CpG, then sees if there are any samples (for a given CpG) that has an extremely high (Q3 + 3*IQR) or extremely low (Q1 - 3*IQR) methylation level, and adds them all up. Before you use this function I would like to preface it by saying that in its current iteration it is extremely slow. Moreover, if you choose to save the outputs (e.g., the identified SEMs for each individual sample), it can take up a lot of disk space. So, just fair warning on that. Either way, here's how to use it:

``` {r}
# detect SEMs
sampleSEMs <- findSEMs(bs, minSamples = 10) # setting sampels to 10 for demo, but >50 is recommended for actual use


# check the output
head(sampleSEMs)

# check to see if there are any differences between groups
t.test(sampleSEMs$log10SEMs ~ factor(pData(bs)$Diagnosis))

```

And that's all there is to it! Similar to above, a Wilcoxon test or glm would probably be more appropriate for your own data. Just keep that in mind!

Good luck out there!

-AM

