<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>MADRID: Methylation Age Determination via Read-Informated Data</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">MADRID: Methylation Age Determination via
Read-Informated Data</h1>
<h3 class="subtitle">Andy Madrid</h3>



<div id="general-usage" class="section level2">
<h2>General Usage</h2>
<p>This is a short introduction to the general functionality of the
madrid package. In short, this package takes a bsseq object, filters
CpGs to only those select few used to determine epigenetic age, impute
any missing values, and then predicts age for each sample.</p>
<p>There are a few more functions, such as guessing the sex of samples,
but those are very much in the beta testing stage.</p>
<p>Let’s start by installing some packages that are needed to install
and run the madrid package.</p>
</div>
<div id="installation" class="section level2">
<h2>Installation</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="do">### Install required packages</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="co"># install.packages(c(&quot;devtools&quot;,&quot;bsseq&quot;,&quot;glmnet&quot;, &quot;mice&quot;, &quot;wateRmelon&quot;, &quot;DunedinPACE&quot;))</span></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="do">### install the madrid package</span></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="co"># install_github(&quot;andymadrid/madrid&quot;)</span></span></code></pre></div>
<p>Now we can load the madrid package.</p>
</div>
<div id="setup" class="section level2">
<h2>Setup</h2>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="co"># load in the madrid pacakge</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(madrid))</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="co">#&gt; Warning: package &#39;GenomicRanges&#39; was built under R version 4.2.2</span></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="co">#&gt; Warning: package &#39;S4Vectors&#39; was built under R version 4.2.2</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="co">#&gt; Warning: package &#39;GenomeInfoDb&#39; was built under R version 4.2.2</span></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a><span class="co">#&gt; Warning: no function found corresponding to methods exports from &#39;BSgenome&#39;</span></span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a><span class="co">#&gt; for: &#39;releaseName&#39;</span></span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a><span class="co">#&gt; Warning: package &#39;limma&#39; was built under R version 4.2.2</span></span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a><span class="co">#&gt; No methods found in package &#39;RSQLite&#39; for request: &#39;dbListFields&#39; when loading &#39;lumi&#39;</span></span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a><span class="co">#&gt; Warning: replacing previous import &#39;GenomicRanges::union&#39; by &#39;dplyr::union&#39;</span></span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a><span class="co">#&gt; when loading &#39;madrid&#39;</span></span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a><span class="co">#&gt; Warning: replacing previous import &#39;GenomicRanges::intersect&#39; by</span></span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a><span class="co">#&gt; &#39;dplyr::intersect&#39; when loading &#39;madrid&#39;</span></span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a><span class="co">#&gt; Warning: replacing previous import &#39;GenomicRanges::setdiff&#39; by &#39;dplyr::setdiff&#39;</span></span>
<span id="cb2-15"><a href="#cb2-15" tabindex="-1"></a><span class="co">#&gt; when loading &#39;madrid&#39;</span></span>
<span id="cb2-16"><a href="#cb2-16" tabindex="-1"></a><span class="co">#&gt; Warning: replacing previous import &#39;bsseq::combine&#39; by &#39;dplyr::combine&#39; when</span></span>
<span id="cb2-17"><a href="#cb2-17" tabindex="-1"></a><span class="co">#&gt; loading &#39;madrid&#39;</span></span>
<span id="cb2-18"><a href="#cb2-18" tabindex="-1"></a><span class="co">#&gt; Warning: replacing previous import &#39;AnnotationHub::hubUrl&#39; by</span></span>
<span id="cb2-19"><a href="#cb2-19" tabindex="-1"></a><span class="co">#&gt; &#39;rtracklayer::hubUrl&#39; when loading &#39;madrid&#39;</span></span></code></pre></div>
<p>Now we can load in a toy dataset.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co"># get toy example</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="fu">data</span>(bs)</span></code></pre></div>
<p>Let us take a look at this toy example.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="co"># summary of bsseq object</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>bs</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="co">#&gt; An object of type &#39;BSseq&#39; with</span></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="co">#&gt;   1335479 methylation loci</span></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a><span class="co">#&gt;   10 samples</span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a><span class="co">#&gt; has not been smoothed</span></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a><span class="co">#&gt; All assays are in-memory</span></span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a></span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a><span class="co"># summary of phenotypic data associated with bsseq object</span></span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a><span class="fu">pData</span>(bs)</span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a><span class="co">#&gt; DataFrame with 10 rows and 3 columns</span></span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a><span class="co">#&gt;                 Age   Diagnosis         Sex</span></span>
<span id="cb4-13"><a href="#cb4-13" tabindex="-1"></a><span class="co">#&gt;           &lt;numeric&gt; &lt;character&gt; &lt;character&gt;</span></span>
<span id="cb4-14"><a href="#cb4-14" tabindex="-1"></a><span class="co">#&gt; Sample_1         31     Control      Female</span></span>
<span id="cb4-15"><a href="#cb4-15" tabindex="-1"></a><span class="co">#&gt; Sample_2         36     Control      Female</span></span>
<span id="cb4-16"><a href="#cb4-16" tabindex="-1"></a><span class="co">#&gt; Sample_3         34     Control      Female</span></span>
<span id="cb4-17"><a href="#cb4-17" tabindex="-1"></a><span class="co">#&gt; Sample_4         53     Control      Female</span></span>
<span id="cb4-18"><a href="#cb4-18" tabindex="-1"></a><span class="co">#&gt; Sample_5         59     Control        Male</span></span>
<span id="cb4-19"><a href="#cb4-19" tabindex="-1"></a><span class="co">#&gt; Sample_6         81        Case      Female</span></span>
<span id="cb4-20"><a href="#cb4-20" tabindex="-1"></a><span class="co">#&gt; Sample_7         56        Case        Male</span></span>
<span id="cb4-21"><a href="#cb4-21" tabindex="-1"></a><span class="co">#&gt; Sample_8         56        Case      Female</span></span>
<span id="cb4-22"><a href="#cb4-22" tabindex="-1"></a><span class="co">#&gt; Sample_9         58        Case        Male</span></span>
<span id="cb4-23"><a href="#cb4-23" tabindex="-1"></a><span class="co">#&gt; Sample_10        56        Case      Female</span></span></code></pre></div>
<p>We see we have 10 samples. Due to size limitations, this bsseq object
only has the CpGs used to predict age, along with CpGs from the sex
chromosomes. However, for your purposes, there is no need to remove any
CpGs from your bsseq object. In fact, that is not advised.</p>
<p>The bsseq object also has some phenotypic data for the samples, which
includes the age, sex, and (pseudo) diagnostic group for each.</p>
</div>
<div id="guess-sex-of-samples" class="section level2">
<h2>Guess sex of samples</h2>
<p>Now we can attempt to guess the sex of the samples.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="co"># guess sample sex from bsseq object</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>guessed_sex <span class="ot">&lt;-</span> <span class="fu">guessSex</span>(bs)</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="co">#&gt; [madrid]: filtering to only CpGs on sex chromosomes</span></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a><span class="co">#&gt; [madrid]: getting medians for each sample</span></span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="co">#&gt; [madrid]: guessing the sex for each sample</span></span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a></span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a><span class="co"># look at the output</span></span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a>guessed_sex</span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a><span class="co">#&gt;    Median_Meth_chrX Median_Meth_chrY predictedSex</span></span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a><span class="co">#&gt; 1         0.8571429        1.0000000       Female</span></span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a><span class="co">#&gt; 2         0.8409091        1.0000000       Female</span></span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a><span class="co">#&gt; 3         0.8636364        1.0000000       Female</span></span>
<span id="cb5-13"><a href="#cb5-13" tabindex="-1"></a><span class="co">#&gt; 4         0.8588957        1.0000000       Female</span></span>
<span id="cb5-14"><a href="#cb5-14" tabindex="-1"></a><span class="co">#&gt; 5         0.9375000        0.8571429         Male</span></span>
<span id="cb5-15"><a href="#cb5-15" tabindex="-1"></a><span class="co">#&gt; 6         0.8709677        1.0000000       Female</span></span>
<span id="cb5-16"><a href="#cb5-16" tabindex="-1"></a><span class="co">#&gt; 7         0.9166667        0.8750000         Male</span></span>
<span id="cb5-17"><a href="#cb5-17" tabindex="-1"></a><span class="co">#&gt; 8         0.8636364        1.0000000       Female</span></span>
<span id="cb5-18"><a href="#cb5-18" tabindex="-1"></a><span class="co">#&gt; 9         0.9285714        0.8500000         Male</span></span>
<span id="cb5-19"><a href="#cb5-19" tabindex="-1"></a><span class="co">#&gt; 10        0.8571429        1.0000000       Female</span></span></code></pre></div>
<p>We see that the guessed sex matches up nicely with the actual sex of
the samples.</p>
</div>
<div id="predict-age" class="section level2">
<h2>Predict age</h2>
<p>Now lets predict the age of the samples</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="co"># predict age from bsseq object</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>predicted_age <span class="ot">&lt;-</span> <span class="fu">predictAge</span>(bs)</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="co">#&gt; [madrid]: getting methylation levels for each sample</span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a><span class="co">#&gt; [madrid]: filtering CpGs to only those predictive of age</span></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="co">#&gt; [madrid]: no missing data, skipping imputation</span></span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a><span class="co">#&gt; [madrid]: predicting age now</span></span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a><span class="co">#&gt; [madrid]: gathering CpGs for array-based clocks now</span></span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a><span class="co">#&gt; Warning in .merge_two_Seqinfo_objects(x, y): Each of the 2 combined objects has sequence levels not in the other:</span></span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a><span class="co">#&gt;   - in &#39;x&#39;: chr7_KI270803v1_alt, chr15_KI270850v1_alt, chr17_KI270909v1_alt, chr4_GL000008v2_random, chr14_GL000009v2_random, chr14_KI270846v1_alt, chr19_KI270938v1_alt, chr22_KI270879v1_alt, chr1_KI270706v1_random, chrUn_KI270742v1, chr1_KI270766v1_alt, chr2_KI270776v1_alt, chr8_KI270821v1_alt, chr17_KI270857v1_alt</span></span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a><span class="co">#&gt;   - in &#39;y&#39;: chrM</span></span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a><span class="co">#&gt;   Make sure to always combine/compare objects based on the same reference</span></span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a><span class="co">#&gt;   genome (use suppressWarnings() to suppress this warning).</span></span>
<span id="cb6-13"><a href="#cb6-13" tabindex="-1"></a><span class="co">#&gt; [madrid]: calculating Dunedin methylation pace now</span></span>
<span id="cb6-14"><a href="#cb6-14" tabindex="-1"></a><span class="co">#&gt; [madrid]: imputing missing data for array-based CpGs now</span></span>
<span id="cb6-15"><a href="#cb6-15" tabindex="-1"></a><span class="co">#&gt; [madrid]: calculating age from array-based clocks now</span></span>
<span id="cb6-16"><a href="#cb6-16" tabindex="-1"></a></span>
<span id="cb6-17"><a href="#cb6-17" tabindex="-1"></a><span class="co"># look at the output</span></span>
<span id="cb6-18"><a href="#cb6-18" tabindex="-1"></a>predicted_age</span>
<span id="cb6-19"><a href="#cb6-19" tabindex="-1"></a><span class="co">#&gt;           MADRID_Age Horvath_Age Hannum_Age PhenoAge SkinBlood_Age  Lin_Age</span></span>
<span id="cb6-20"><a href="#cb6-20" tabindex="-1"></a><span class="co">#&gt; Sample_1    34.48592    34.60565 -4.1621911   60.664      8.295322 12.21698</span></span>
<span id="cb6-21"><a href="#cb6-21" tabindex="-1"></a><span class="co">#&gt; Sample_2    40.80927    34.60565 -3.4333561   60.664      8.665987 12.21698</span></span>
<span id="cb6-22"><a href="#cb6-22" tabindex="-1"></a><span class="co">#&gt; Sample_3    36.01735    34.60565 -6.1512065   60.664      8.407280 12.21698</span></span>
<span id="cb6-23"><a href="#cb6-23" tabindex="-1"></a><span class="co">#&gt; Sample_4    54.91187    34.60565 -2.6394282   60.664      8.694256 12.21698</span></span>
<span id="cb6-24"><a href="#cb6-24" tabindex="-1"></a><span class="co">#&gt; Sample_5    56.54770    34.60565 -0.8991330   60.664     10.024155 12.21698</span></span>
<span id="cb6-25"><a href="#cb6-25" tabindex="-1"></a><span class="co">#&gt; Sample_6    73.48403    34.60565 -2.3894243   60.664      9.462471 12.21698</span></span>
<span id="cb6-26"><a href="#cb6-26" tabindex="-1"></a><span class="co">#&gt; Sample_7    53.69849    34.60565 -1.8131179   60.664      9.663648 12.21698</span></span>
<span id="cb6-27"><a href="#cb6-27" tabindex="-1"></a><span class="co">#&gt; Sample_8    58.76004    34.60565  0.3265283   60.664      8.961952 12.21698</span></span>
<span id="cb6-28"><a href="#cb6-28" tabindex="-1"></a><span class="co">#&gt; Sample_9    57.31309    34.60565 -1.9617247   60.664      9.503491 12.21698</span></span>
<span id="cb6-29"><a href="#cb6-29" tabindex="-1"></a><span class="co">#&gt; Sample_10   52.35543    34.60565 -4.1841212   60.664      9.174092 12.21698</span></span>
<span id="cb6-30"><a href="#cb6-30" tabindex="-1"></a><span class="co">#&gt;           DunedinPACE</span></span>
<span id="cb6-31"><a href="#cb6-31" tabindex="-1"></a><span class="co">#&gt; Sample_1           NA</span></span>
<span id="cb6-32"><a href="#cb6-32" tabindex="-1"></a><span class="co">#&gt; Sample_2           NA</span></span>
<span id="cb6-33"><a href="#cb6-33" tabindex="-1"></a><span class="co">#&gt; Sample_3           NA</span></span>
<span id="cb6-34"><a href="#cb6-34" tabindex="-1"></a><span class="co">#&gt; Sample_4           NA</span></span>
<span id="cb6-35"><a href="#cb6-35" tabindex="-1"></a><span class="co">#&gt; Sample_5           NA</span></span>
<span id="cb6-36"><a href="#cb6-36" tabindex="-1"></a><span class="co">#&gt; Sample_6           NA</span></span>
<span id="cb6-37"><a href="#cb6-37" tabindex="-1"></a><span class="co">#&gt; Sample_7           NA</span></span>
<span id="cb6-38"><a href="#cb6-38" tabindex="-1"></a><span class="co">#&gt; Sample_8           NA</span></span>
<span id="cb6-39"><a href="#cb6-39" tabindex="-1"></a><span class="co">#&gt; Sample_9           NA</span></span>
<span id="cb6-40"><a href="#cb6-40" tabindex="-1"></a><span class="co">#&gt; Sample_10          NA</span></span></code></pre></div>
<p>This outputs the predicted ages for each sample. We can also see how
well they correlate with the actual ages, if that information is readily
available.</p>
<p>Please note, by default the predictAge() function will estimate age
using array-based clocks, as well as the sequencing-based clock. To do
this, it leans heavily on two other packages, wateRmelon and
DunedinPACE. At its core, for this portion of the function, it really
just formats the methylation matrix accordingly and filters CpGs to only
those on the arrays. In this toy example most of the ages look off due
to the fact that most CpGs are missing because of size limitations. That
being said, in your own data, estimates should (hopefully) be more
accurate.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="co"># assess the correlation between actual and estimated age</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>corMADRID <span class="ot">&lt;-</span> <span class="fu">cor</span>(predicted_age<span class="sc">$</span>MADRID_Age, <span class="fu">pData</span>(bs)<span class="sc">$</span>Age)</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a>corMADRID</span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a><span class="co">#&gt; [1] 0.9853733</span></span></code></pre></div>
<p>We see that we have a fairly nice correlation between predicted and
actual age. And that’s really all that there is to it. You can take
these results and see if there are any signficiant differences between
the estimated and actual age of samples between groups.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="co"># test for differences (accelerated/decelerated) in actual and estimated age</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>predicted_age<span class="sc">$</span>difference <span class="ot">&lt;-</span> predicted_age<span class="sc">$</span>MADRID_Age <span class="sc">-</span> <span class="fu">pData</span>(bs)<span class="sc">$</span>Age</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a><span class="fu">t.test</span>(predicted_age<span class="sc">$</span>difference <span class="sc">~</span> <span class="fu">factor</span>(<span class="fu">pData</span>(bs)<span class="sc">$</span>Diagnosis))</span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a><span class="co">#&gt;  Welch Two Sample t-test</span></span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a><span class="co">#&gt; data:  predicted_age$difference by factor(pData(bs)$Diagnosis)</span></span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a><span class="co">#&gt; t = -2.0276, df = 7.2851, p-value = 0.08062</span></span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a><span class="co">#&gt; alternative hypothesis: true difference in means between group Case and group Control is not equal to 0</span></span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a><span class="co">#&gt; 95 percent confidence interval:</span></span>
<span id="cb8-12"><a href="#cb8-12" tabindex="-1"></a><span class="co">#&gt;  -9.1290538  0.6646447</span></span>
<span id="cb8-13"><a href="#cb8-13" tabindex="-1"></a><span class="co">#&gt; sample estimates:</span></span>
<span id="cb8-14"><a href="#cb8-14" tabindex="-1"></a><span class="co">#&gt;    mean in group Case mean in group Control </span></span>
<span id="cb8-15"><a href="#cb8-15" tabindex="-1"></a><span class="co">#&gt;             -2.277782              1.954423</span></span></code></pre></div>
<p>Some people prefer to run the test on the residuals, rather than on
the direct differences in age themselves. That would look something like
this…</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="co"># test for differences in the resduals between actual and estimated age</span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>predicted_age<span class="sc">$</span>residuals <span class="ot">&lt;-</span> <span class="fu">resid</span>(<span class="fu">lm</span>(predicted_age<span class="sc">$</span>MADRID_Age <span class="sc">~</span> <span class="fu">pData</span>(bs)<span class="sc">$</span>Age))</span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a><span class="fu">t.test</span>(predicted_age<span class="sc">$</span>residuals <span class="sc">~</span> <span class="fu">factor</span>(<span class="fu">pData</span>(bs)<span class="sc">$</span>Diagnosis))</span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a><span class="co">#&gt;  Welch Two Sample t-test</span></span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a><span class="co">#&gt; data:  predicted_age$residuals by factor(pData(bs)$Diagnosis)</span></span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a><span class="co">#&gt; t = -0.089266, df = 7.1588, p-value = 0.9313</span></span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a><span class="co">#&gt; alternative hypothesis: true difference in means between group Case and group Control is not equal to 0</span></span>
<span id="cb9-10"><a href="#cb9-10" tabindex="-1"></a><span class="co">#&gt; 95 percent confidence interval:</span></span>
<span id="cb9-11"><a href="#cb9-11" tabindex="-1"></a><span class="co">#&gt;  -3.296327  3.055463</span></span>
<span id="cb9-12"><a href="#cb9-12" tabindex="-1"></a><span class="co">#&gt; sample estimates:</span></span>
<span id="cb9-13"><a href="#cb9-13" tabindex="-1"></a><span class="co">#&gt;    mean in group Case mean in group Control </span></span>
<span id="cb9-14"><a href="#cb9-14" tabindex="-1"></a><span class="co">#&gt;           -0.06021588            0.06021588</span></span></code></pre></div>
<p>I’d like to note that a Kruskall-Wallis test or something like a
generalized linear model that adjusts for known covariates would
probably be more approporaite test to use, but for the sake of
demonstration a simple t-test is utilized here.</p>
<p>That being said, in this toy example there is no signficant
difference (accelerated or decelerated) in predicted age between groups.
Hopefully in your data there will be, though!</p>
<p>If you’re using blood samples, we can also estiamte the cell-type
proportions from your sequenced data. The function estimateCellCounts()
leans heavily upon the R package meffil to perform these calculations.
In testing, I had some samples that were run on the EPIC microarray and
were also sequenced and found that this method produced results that
were more correlated with each other than other methods that were
developed for sequence data (e.g., methylcc).</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="co"># estimate cell-type proportions from blood samples</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="co"># cellProps &lt;- estimateCellCounts(bs)</span></span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a></span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a><span class="co"># check the estimates</span></span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a><span class="co"># head(cellProps)</span></span></code></pre></div>
<p>Please note, the above code is not actually run since the bsseq
object we have here for the demonstration does not have the sufficient
CpGs needed estimate cell-type proportions and would throw out an error.
But, assuming your bsseq object has all the CpGs, that shouldn’t be an
issue for you!</p>
<p>When working with human data, people often align to one of two genome
builds - either hg19 or hg38. Nowadays, people pretty much just align to
hg38. But, on the off chance that you aligned to hg19 or are using old
data that was aligned to hg19, or if you want to convert your hg38
coordinates to hg19 for whatever reason, I’ve implemented a function to
do just that. See below:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="co"># lift coordinates from hg38 to hg19</span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>bsLifted <span class="ot">&lt;-</span> <span class="fu">liftBuild</span>(bs, <span class="at">current =</span> <span class="st">&quot;hg38&quot;</span>, <span class="at">new =</span> <span class="st">&quot;hg19&quot;</span>)</span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a><span class="co">#&gt; [madrid]: Will lift your bsseq object from hg38 to hg19</span></span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a><span class="co">#&gt; snapshotDate(): 2022-04-25</span></span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a><span class="co">#&gt; loading from cache</span></span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a><span class="co">#&gt; [madrid]: Lifting over resulted in a loss of 68787 CpGs</span></span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a></span>
<span id="cb11-8"><a href="#cb11-8" tabindex="-1"></a><span class="co"># check the lifted coordinates</span></span>
<span id="cb11-9"><a href="#cb11-9" tabindex="-1"></a><span class="co">#bsLifted</span></span></code></pre></div>
<p>While epigenetic age is the main function of this package, I’ve also
added in a function to identify stochastic epigenetic mutations (SEMs).
For more information on SEMs please refer to Gentilini et al (2015)
Aging and Markov et al (2024) GeroScience.</p>
<p>Essentially, this function takes your data, identifies the
interquartile range (IQR) for each CpG, then sees if there are any
samples (for a given CpG) that has an extremely high (Q3 + 3<em>IQR) or
extremely low (Q1 - 3</em>IQR) methylation level, and adds them all up.
Before you use this function I would like to preface it by saying that
in its current iteration it is extremely slow. Moreover, if you choose
to save the outputs (e.g., the identified SEMs for each individual
sample), it can take up a lot of disk space. So, just fair warning on
that. Either way, here’s how to use it:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="co"># detect SEMs</span></span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a><span class="co"># sampleSEMs &lt;- findSEMs(bs, minSamples = 10) # setting samples to 10 for demo, but &gt;50 is recommended for actual use</span></span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a></span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a></span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a><span class="co"># check the output</span></span>
<span id="cb12-6"><a href="#cb12-6" tabindex="-1"></a><span class="fu">head</span>(sampleSEMs)</span>
<span id="cb12-7"><a href="#cb12-7" tabindex="-1"></a><span class="co">#&gt;          Hyper_SEMs Hypo_SEMs Total_SEMs log10SEMs</span></span>
<span id="cb12-8"><a href="#cb12-8" tabindex="-1"></a><span class="co">#&gt; Sample_1        376      5565       5941  3.773860</span></span>
<span id="cb12-9"><a href="#cb12-9" tabindex="-1"></a><span class="co">#&gt; Sample_2        696      9623      10319  4.013638</span></span>
<span id="cb12-10"><a href="#cb12-10" tabindex="-1"></a><span class="co">#&gt; Sample_3        512      4501       5013  3.700098</span></span>
<span id="cb12-11"><a href="#cb12-11" tabindex="-1"></a><span class="co">#&gt; Sample_4        530      9227       9757  3.989316</span></span>
<span id="cb12-12"><a href="#cb12-12" tabindex="-1"></a><span class="co">#&gt; Sample_5       6516      5855      12371  4.092405</span></span>
<span id="cb12-13"><a href="#cb12-13" tabindex="-1"></a><span class="co">#&gt; Sample_6        583      5027       5610  3.748963</span></span>
<span id="cb12-14"><a href="#cb12-14" tabindex="-1"></a></span>
<span id="cb12-15"><a href="#cb12-15" tabindex="-1"></a><span class="co"># check to see if there are any differences between groups</span></span>
<span id="cb12-16"><a href="#cb12-16" tabindex="-1"></a><span class="fu">t.test</span>(sampleSEMs<span class="sc">$</span>log10SEMs <span class="sc">~</span> <span class="fu">factor</span>(<span class="fu">pData</span>(bs)<span class="sc">$</span>Diagnosis))</span>
<span id="cb12-17"><a href="#cb12-17" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb12-18"><a href="#cb12-18" tabindex="-1"></a><span class="co">#&gt;  Welch Two Sample t-test</span></span>
<span id="cb12-19"><a href="#cb12-19" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb12-20"><a href="#cb12-20" tabindex="-1"></a><span class="co">#&gt; data:  sampleSEMs$log10SEMs by factor(pData(bs)$Diagnosis)</span></span>
<span id="cb12-21"><a href="#cb12-21" tabindex="-1"></a><span class="co">#&gt; t = -0.72429, df = 6.2028, p-value = 0.4953</span></span>
<span id="cb12-22"><a href="#cb12-22" tabindex="-1"></a><span class="co">#&gt; alternative hypothesis: true difference in means between group Case and group Control is not equal to 0</span></span>
<span id="cb12-23"><a href="#cb12-23" tabindex="-1"></a><span class="co">#&gt; 95 percent confidence interval:</span></span>
<span id="cb12-24"><a href="#cb12-24" tabindex="-1"></a><span class="co">#&gt;  -0.2699336  0.1458769</span></span>
<span id="cb12-25"><a href="#cb12-25" tabindex="-1"></a><span class="co">#&gt; sample estimates:</span></span>
<span id="cb12-26"><a href="#cb12-26" tabindex="-1"></a><span class="co">#&gt;    mean in group Case mean in group Control </span></span>
<span id="cb12-27"><a href="#cb12-27" tabindex="-1"></a><span class="co">#&gt;              3.851835              3.913863</span></span></code></pre></div>
<p>You’ll notice that a portion of the script is commented out. That was
done to speed things up. I will be honest and say that this function is
pretty slow. If there’s a future need for it, I can work on speeding it
up. But for now, I’ll leave it as is.</p>
<p>And that’s all there is to it! Similar to above, a Kruskall-Wallis
test or glm would probably be more appropriate for your own data. Just
keep that in mind!</p>
<p>Good luck out there!</p>
<p>-AM</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
